consensus/parlia/parlia.go

```diff
@@ -713,25 +713,40 @@ func (p *Parlia) snapshot(chain consensus.ChainHeaderReader, number uint64, hash
 			}
 		}
 
-		// If we're at the genesis, snapshot the initial state.
-		if number == 0 {
-			checkpoint := chain.GetHeaderByNumber(number)
-			if checkpoint != nil {
-				// get checkpoint data
-				hash := checkpoint.Hash()
-
+		// If we're at the genesis, snapshot the initial state. Alternatively if we have
+		// piled up more headers than allowed to be reorged (chain reinit from a freezer),
+		// consider the checkpoint trusted and snapshot it.
+		// even BEP-341 enabled, an offset `p.config.Epoch/2` can ensure getting the right validators.
+		if number == 0 || (number%p.config.Epoch == p.config.Epoch/2 && (len(headers) > int(params.FullImmutabilityThreshold))) {
```

Comment:
Is it ok to parse validators in `p.config.Epoch/2`? maybe `p.config.Epoch-1` is a good choice, because we do not know the validators length.